Generic Protocol breaks mypy when attrs class

Please provide more information to help us understand the issue:

* Are you reporting a bug, or opening a feature request?
Bug

* Please insert below the code you are checking with mypy,
  or a mock-up repro if the source is private. We would appreciate
  if you try to simplify your case to a minimal repro.
```
from abc import abstractmethod
from typing import TypeVar

from attr import attrs
from typing_extensions import Protocol

T = TypeVar("T")


@attrs(auto_attribs=True)
class Test(Protocol[T]):
    a: int

    @abstractmethod
    def transition_to(self, t: T):
        raise NotImplementedError
```

* What is the actual behavior/output?
mypy crashes
* What is the behavior/output you expect?
```
test.py:20:1: error: Invariant type variable 'T' used in protocol where contravariant one is expected
test.py:24:5: error: Function is missing a return type annotation
```

* What are the versions of mypy and Python you are using?
  Do you see the same issue after installing mypy from Git master?
Using 701, still exists with master

* What are the mypy flags you are using? (For example --strict-optional)
No config

* If mypy crashed with a traceback, please paste
  the full traceback below.

```
test.py:24:5: error: Function is missing a return type annotation
./test.py:19: error: INTERNAL ERROR -- please report a bug at https://github.com/python/mypy/issues version: 0.701
Traceback (most recent call last):
  File "/Users/cahurst/.local/share/virtualenvs/automation_engine-fTGkomR2/bin/mypy", line 10, in <module>
    sys.exit(console_entry())
  File "mypy/checker.py", line 395, in accept
  File "mypy/nodes.py", line 896, in accept
  File "mypy/checker.py", line 1528, in visit_class_def__StatementVisitor_glue
  File "mypy/checker.py", line 1568, in visit_class_def
  File "mypy/checker.py", line 1588, in check_protocol_variance
  File "mypy/subtypes.py", line 95, in is_subtype
  File "mypy/types.py", line 625, in accept
  File "mypy/subtypes.py", line 185, in visit_instance__TypeVisitor_glue
  File "mypy/subtypes.py", line 218, in visit_instance
  File "mypy/subtypes.py", line 482, in is_protocol_implementation
AssertionError:
./test.py:19: : note: use --pdb to drop into pdb
```


Removing either the Generic aspect, or removing attrs from the class makes everything work as expected. It only occurs when both are in play at the same time.

--- a/mypy/nodes.py
+++ b/mypy/nodes.py
@@ -2906,7 +2906,10 @@ class TypeInfo(SymbolNode):
         assert self.mro, "This property can be only accessed after MRO is (re-)calculated"
         for base in self.mro[:-1]:  # we skip "object" since everyone implements it
             if base.is_protocol:
-                for name in base.names:
+                for name, node in base.names.items():
+                    if isinstance(node.node, (TypeAlias, TypeVarExpr)):
+                        # These are auxiliary definitions (and type aliases are prohibited).
+                        continue
                     members.add(name)
         return sorted(list(members))
 

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -3269,6 +3269,12 @@ class SemanticAnalyzer(
         current_node = existing.node if existing else alias_node
         assert isinstance(current_node, TypeAlias)
         self.disable_invalid_recursive_aliases(s, current_node)
+        if self.is_class_scope():
+            assert self.type is not None
+            if self.type.is_protocol:
+                self.fail("Type aliases are prohibited in protocol bodies", s)
+                if not lvalue.name[0].isupper():
+                    self.note("Use variable annotation syntax to define protocol members", s)
         return True
 
     def disable_invalid_recursive_aliases(

--- a/test-data/unit/check-protocols.test
+++ b/test-data/unit/check-protocols.test
@@ -3546,3 +3546,34 @@ S = TypeVar("S")
 def test(arg: P[S]) -> S: ...
 b: Type[B]
 reveal_type(test(b))  # N: Revealed type is "__main__.B"
+
+[case testTypeAliasInProtocolBody]
+from typing import Protocol, List
+
+class P(Protocol):
+    x = List[str]  # E: Type aliases are prohibited in protocol bodies \
+                   # N: Use variable annotation syntax to define protocol members
+
+class C:
+    x: int
+def foo(x: P) -> None: ...
+foo(C())  # No extra error here
+[builtins fixtures/list.pyi]
+
+[case testTypeVarInProtocolBody]
+from typing import Protocol, TypeVar
+
+class C(Protocol):
+    T = TypeVar('T')
+    def __call__(self, t: T) -> T: ...
+
+def f_bad(t: int) -> int:
+    return t
+
+S = TypeVar("S")
+def f_good(t: S) -> S:
+    return t
+
+g: C = f_bad  # E: Incompatible types in assignment (expression has type "Callable[[int], int]", variable has type "C") \
+              # N: "C.__call__" has type "Callable[[Arg(T, 't')], T]"
+g = f_good  # OK

